Assignment Link: https://github.com/DataTalksClub/mlops-zoomcamp/blob/main/02-experiment-tracking/homework.md

# Q1. Install MLflow

In [13]:
! mlflow --version

mlflow, version 1.26.0


Answer-1: MLflow version is 1.26.0

# Q2. Download and preprocess the data

In [23]:
! python3.9 ./preprocess_data.py --raw_data_path . --dest_path ./output

Asnwer-2: There are 4 files in the output folder.

# Q3. Train a model with autolog

In [19]:
! python3.9 ./train.py --data_path ./output

2022/05/30 12:21:19 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of sklearn. If you encounter errors during autologging, try upgrading / downgrading sklearn to a supported version, or try upgrading MLflow.
2022/05/30 12:21:19 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/murathansaygili/Desktop/MLOpsZoomCamp/venv/lib/python3.9/site-packages/setuptools/distutils_patch.py:25: UserWarning: Distutils was imported before Setuptools. This usage is discouraged and may exhibit undesirable behaviors or errors. Please use Setuptools' objects directly or at least import Setuptools first."


Answer-3: UI shows that there are 17 parameters that were automatically logged.

# Q4. Launch the tracking server locally

In [ ]:
# ! mlflow ui --backend-store-uri sqlite:///mlflow.db --default-artifact-root ./artifacts

Answer-4: deafult-artifact-root should be specified.

# Q5. Tune the hyperparameters of the model

In [20]:
! python3.9 ./hpo.py --data_path ./output

2022/05/30 12:37:56 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.
100%|█████████| 50/50 [05:21<00:00,  6.42s/trial, best loss: 6.6284257482044735]


Answer-5: The best validation RMSE=6.628

# Q6. Promote the best model to the model registry

In [25]:
! python3.9 ./register_model.py --data_path ./output

2022/05/30 19:09:29 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of sklearn. If you encounter errors during autologging, try upgrading / downgrading sklearn to a supported version, or try upgrading MLflow.
2022/05/30 19:09:29 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/murathansaygili/Desktop/MLOpsZoomCamp/venv/lib/python3.9/site-packages/setuptools/distutils_patch.py:25: UserWarning: Distutils was imported before Setuptools. This usage is discouraged and may exhibit undesirable behaviors or errors. Please use Setuptools' objects directly or at least import Setuptools first."
d1a263525d844346969511e5febc0432
Registered model 'the_best_model' already exists. Creating a new version of this model...
2022/05/30 19:10:42 INFO mlflow.tracking._model_registry.client: Waiting up to 300 seconds for model version to finish creation.                     Model name: the_best_model, version 3
Created version '3' of model 

Answer-6: The best test RMSE=6.549